AudioKitSynthOne/DSP/Kernel/S1DSPCompressor.hpp

```diff
@@ -0,0 +1,85 @@
+//
+//  S1DSPCompressor.hpp
+//  AudioKitSynthOne
+//
+//  Created by Matthias Frick on 11/03/2019.
+//  Copyright © 2019 AudioKit. All rights reserved.
+//
+
+#import "AudioKit/AKSoundpipeKernel.hpp"
+#import "S1Parameter.h"
+#include <iostream>
+
+#ifndef S1DSPCompressor_h
+#define S1DSPCompressor_h
+using DSPParameters = std::array<float, S1Parameter::S1ParameterCount>;
+
+template<int RatioP, int ThresholdP, int AttP, int RelP, int MakeupP = -1>
+struct S1Compressor {
+    S1Compressor(DSPParameters* params) {
+        
+    };
+
+    void compute() {
+        
+    };
+
+    S1Compressor() = delete; // This should be RAII, Don't allow compressor without kernel
+    S1Compressor(sp_data* sp) :
+        mSp(sp)
+    {
+        sp_compressor_create(&mCompressorR);
+        sp_compressor_create(&mCompressorL);
+        sp_compressor_init(mSp, mCompressorR);
+        sp_compressor_init(mSp, mCompressorL);
+    };
+
+    void update(const DSPParameters &params) {
+        configure(mCompressorR, params);
+        configure(mCompressorL, params);
+    }
+
+    void compute(float &inL, float &inR, float &outL, float &outR) {
+        compute(mCompressorR, inR, outR);
+        compute(mCompressorL, inL, outL);
+        if (MakeupP != -1) {
+            outR *= makeUpGain;
+            outL *= makeUpGain;
+        }
+    }
+
+    ~S1Compressor() {
```

Comment:
right now, copying/moving of S1Compressor is allowed, leading to dangling pointers to the sp_compressor instances and double destroying. For clean RAII, we should either disable or define copy and move constructors (rule of five) _or_ preferably, simply wrap the sp_compressor instances in unique_ptrs with custom deletion so we can get rid of the destructor (rule of zero).

Reply:
Copying/Moving is implicitly deleted, but should make that explicit. I think deleting them is better in the long run, because they hold a reference to the `sp` (Soundpipe Kernel) and when they get moved or copied around, changes are high that reference might be invalid too.

And good call on the use for `unique_ptr` with custom deleters. I'll push a fixup for that later 👍

Reply:
Ah i see why moving is implicitly deleted, but to my understanding copying would still be possible. What am I missing?

Reply:
Yeah, sorry. I was only refering to moving. Either way, I'l sort those out later. Mind if I ping you for another review? :)

Reply:
Yeah, sorry. I was only refering to moving. Either way, I'l sort those out later. Mind if I ping you for another review? :)

Reply:
no problem 👍

Reply:
@tusueso might taking a look at 
https://github.com/AudioKit/AudioKitSynthOne/pull/56/commits/cf8ab229b39b11302967eb90ac855eb008849423 ?